### ISSUE TITLE
C++ library

### ISSUE THREAD

**Issue by bdon** [flatgeobuf/flatgeobuf#125]
Hi, I'm experimenting with writing a direct raw OSM to FGB converter (https://github.com/bdon/osm2fgb , maybe something like this exists already) and have a few questions about FGB:

* What's the recommended way to include the C++ code as a dependency? Is there an existing project that does this I could copy? FlatBuffers offers a CMake project that can be included as a subdirectory inside CMake, so if nothing else I would suggest that as a possibility and can look into adding that...

* It would be useful to do stateful streaming serialization; seems like the way to stream it now is via event-based lambdas, but it would be tricky to make this work with another callback-based library like osmium (https://github.com/bdon/osm2fgb/blob/master/src/main.cpp#L31); something like

```c++
flatgeobuf::Serializer s{outfile};
s.write(buff_ptr, size);
s.finalize(true); // create indexes
```

* I noticed that a fixed column schema is enforced by the serializer by peeking at the first feature, is this a design constraint of the C++ implementation, or is "schema less" key-value like in OSM possible? 

The potential use case here:  use thematic data in QGIS from OSM for building polygons and landuse polygons. To make this work with FGB it sounds like I need to pre-define columns (tags) for each theme, and also keep each theme in a separate FGB file (advantage Geopackage = multiple "layers" in one file)

**Comment by pka** [comment]
:+1: from my side for your project! Since I'm on the Rust side, I can't help you with C++ unfortunately. Did you check what [ogr2ogr](https://gdal.org/drivers/vector/flatgeobuf.html) produces from a OSM PBF?

**Comment by bdon** [comment]
@pka good idea :) just played with it now, some observations of how the `ogr2ogr -f flatgeobuf` works:

* Each of the defined layers in osmconf.ini is exposed by the OGR driver as a "layer" and needs to be exported in a separate pass to FGB
* from my experience the OGR OSM reader is useful for basic use cases but once you want to work with huge PBFs and create complex features from relations/multipolygons than Osmium is one of the only choices, but my knowledge may be out of date on this
* looks like the OGR writer can write NULL values for keys but using the C++ library directly doesn't seem to allow this? https://github.com/flatgeobuf/flatgeobuf/blob/master/src/cpp/geojson.h#L105
* In practice the internal data model of FGB seems to allow 1. mixed geometry types in a single layer and 2. arbitrary keys/values for feature properties. But it seems that the common readers (which aside from GDAL/OGR) will assume a fixed column schema and homogenous geometry types within one file? If this is an official constraint of the specification, I think that makes sense as a direct substitute for Shapefile, but does mean less flexibility than geojson

**Comment by bjornharrtell** [comment]
Hi @bdon,

Nice to see some interest in this department! The C++ reference code is not very mature and lacks documentation. I would like to improve that situation but C++ is not my primary language so bear with me.

That said, Flatbuffers and code gen for C++ is headers only and is essentially enough to get started with serialization/deserialization so there is no need for a library per se. The headers are shared and identical with implementation in GDAL. The index implementation is also shared with GDAL and I'm considering making it header only too. This is the low level parts and as GDAL demonstrate is enough to fully implement the format.

The other part of the reference C++ implementation is as you have seen very naive. It also depends on some mapbox header parts to be able to convert to GeoJSON for simple roundtrip tests. It is not intended as a complete and ideal implementation and should be more clearly separated from the other parts. It is essentially only supporting code for the test suite.

Initial versions of the format constrained the schema to be as you say, shapefile like. I want the format to be as simple as possible. But I realized I could make it less restrictive with very little complexity added so what became the stable spec is more flexible and it's perfectly ok to use and intended to make it possible to be more or less 1-1 with GeoJSON. What makes most sense to do in your case I'm not sure.

As for I/O and streaming there is only the lambda stuff for the index part and nothing else at this time. 
It would of course be nice to provide a more general purpose and complete high level C++ API  but that will require a rather large effort which I'm uncertain if I ever will be able to do alone. Improvements/additions are welcome.

**Comment by bdon** [comment]
@bjornharrtell thank you, I have been creating an intermediate format - WKT CSV or WKB in SQLite - and then using ogr2ogr -f flatgeobuf, and this is sufficient for my needs.

**Comment by district10** [comment]
bdon: 

> I noticed that a fixed column schema is enforced by the serializer by peeking at the first feature, is this a design constraint of the C++ implementation, or is "schema less" key-value like in OSM possible?

Hi, @bjornharrtell , you said "the stable spec is more flexible and it's perfectly ok to use and intended to make it possible to be more or less 1-1 with GeoJSON". It that what we have now? I checked c++ code and typescript code, they all `introspectHeaderMeta` with first feature. Is flatgeobuf a 100% lossless alternative to GeoJSON (a little precision loss like geobuf is okay)?

And what's the reference implementation? Is there a binary I can use to convert arbitrary geojson file to flatgeobuf and then round trip back to geojson?

**Comment by bjornharrtell** [comment]
@district10 the spec allows per feature schema but it's not implemented in most (all?) of the reference implementations. However I think it's implemented in GDAL if I recall correctly which does have a binary distribution and tool to convert (ogr2ogr), you should try that.

I've also made an experimental cli tool in rust here https://github.com/flatgeobuf/flatgeobuf/tree/master/src/rust/fgbutil but it's rather rough and I'm not sure it can/will handle per feature schema.

**Comment by district10** [comment]
@bjornharrtell Thanks for quick reply. I'll try ogr2ogr and come back later (in a few days maybe).

**Comment by district10** [comment]
I encountered some difficulties while working with GDAL. It appears to consistently require a specific schema, which prevents me from converting arbitrary GeoJSON files into other formats that GDAL supports.

My requirements are straightforward: 1) the ability to work with GeoJSON/Geobuf data, 2) the capability for random access to the ith feature within the dataset, and 3) the use of a PackedRTree for efficient bounding box queries. To address these challenges, I developed an indexing solution that parses Geobuf files and identifies each feature by its data range (defined by offset and length), as well as implements a PackedRTree for spatial indexing. This method is just as flexible and schema-free as GeoJSON itself.

proto: https://github.com/cubao/geobuf-cpp/blob/dev/geobuf_index.proto
tests: https://github.com/cubao/geobuf-cpp/blob/b06e682bae6f04ad9147659edbd8eecb9dfe2d3f/tests/test_geobuf.py#L2017-L2030
